File: src/main/cpp/2022/04/AoC2022_04.cpp
#include <assert.h>

#include <algorithm>
#include <string>
#include <vector>

#include "../../aoc/aoc.hpp"
#include "../../aocd/aocd.hpp"

using namespace std;
using namespace aoc;

int solve(const vector<string>& input,
          const function<bool(const Range&, const Range&)>& test) {
    return count_if(input.begin(), input.end(), [&test](const string& line) {
        const vector<int>& nums = getNumbers(line);
        const Range range1 = Range::rangeClosed(nums[0], nums[1]);
        const Range range2 = Range::rangeClosed(nums[2], nums[3]);
        return test(range1, range2);
    });
}

int part1(const vector<string>& input) {
    return solve(input, [](const Range& range1, const Range& range2) {
        return range1.contains(range2) || range2.contains(range1);
    });
}

int part2(const vector<string>& input) {
    return solve(input, [](const Range& range1, const Range& range2) {
        return range1.isOverlappedBy(range2);
    });
}

// clang-format off
const vector<string> TEST = {
"2-4,6-8",
"2-3,4-5",
"5-7,7-9",
"2-8,3-7",
"6-6,4-6",
"2-6,4-8",
};
// clang-format on

void samples() {
    assert(part1(TEST) == 2);
    assert(part2(TEST) == 4);
}

MAIN(2022, 4)

File: src/main/java/AoC2022_04.java
import java.util.List;
import java.util.function.BiPredicate;
import java.util.regex.MatchResult;
import java.util.regex.Pattern;

import com.github.pareronia.aoc.Range;
import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

public class AoC2022_04 extends AoCBase {
    
    private static final Pattern REGEX = Pattern.compile("[0-9]+");
    
    private final List<String> input;
    
    private AoC2022_04(final List<String> input, final boolean debug) {
        super(debug);
        this.input = input;
    }
    
    public static final AoC2022_04 create(final List<String> input) {
        return new AoC2022_04(input, false);
    }

    public static final AoC2022_04 createDebug(final List<String> input) {
        return new AoC2022_04(input, true);
    }
    
    private final int[] numbers(final String string) {
        return REGEX.matcher(string).results()
                .map(MatchResult::group)
                .mapToInt(Integer::parseInt)
                .toArray();
    }
    
    private int solve(final BiPredicate<Range<Integer>, Range<Integer>> predicate) {
        return (int) this.input.stream()
            .map(this::numbers)
            .filter(nums -> predicate.test(
                    Range.between(nums[0], nums[1]),
                    Range.between(nums[2], nums[3])))
            .count();
    }
    
    @Override
    public Integer solvePart1() {
        return solve((range1, range2) ->
                range1.containsRange(range2) || range2.containsRange(range1));
    }

    @Override
    public Integer solvePart2() {
        return solve((range1, range2) -> range1.isOverlappedBy(range2));
    }

    public static void main(final String[] args) throws Exception {
        assert AoC2022_04.createDebug(TEST).solvePart1() == 2;
        assert AoC2022_04.createDebug(TEST).solvePart2() == 4;

        final Puzzle puzzle = Aocd.puzzle(2022, 4);
        final List<String> inputData = puzzle.getInputData();
        puzzle.check(
            () -> lap("Part 1", AoC2022_04.create(inputData)::solvePart1),
            () -> lap("Part 2", AoC2022_04.create(inputData)::solvePart2)
        );
    }

    private static final List<String> TEST = splitLines(
        "2-4,6-8\r\n" +
        "2-3,4-5\r\n" +
        "5-7,7-9\r\n" +
        "2-8,3-7\r\n" +
        "6-6,4-6\r\n" +
        "2-6,4-8"
    );
}
